jupyterhub_config.py

```diff
@@ -220,48 +191,126 @@ def finalize_user_content(spawner):
 # TLS config
 c.JupyterHub.port = 80
 
+# Configure idle culler service (stop idle user containers)
+# This used to be a Python script we had to grab directly, but is now
+# fully integrated as a (non-admin) hub service
+c.JupyterHub.services = [
+    {
+        'name': 'idle-culler',
+        'command': [sys.executable, '-m', 'jupyterhub_idle_culler', '--timeout=3600'],
+    }
+]
+c.JupyterHub.load_roles = [
+    {
+        'name': 'list-and-cull', # name the role
+        'services': [
+            'idle-culler', # assign the service to this role
+        ],
+        'scopes': [
+            # declare what permissions the service should have
+            'list:users', # list users
+            'read:users:activity', # read user last-activity
+            'admin:servers', # start/stop servers
+        ],
+    }
+]
+
+# JH Authentication Config
+c.JupyterHub.authenticator_class = 'oauthenticator.auth0.Auth0OAuthenticator'
+
 # Production authenticator
 c.Auth0OAuthenticator.client_id = os.environ['AUTH0_CLIENT_ID']
 c.Auth0OAuthenticator.client_secret = os.environ['AUTH0_CLIENT_SECRET']
 c.Auth0OAuthenticator.oauth_callback_url = os.environ['AUTH0_CALLBACK_URL']
-c.JupyterHub.authenticator_class = 'oauthenticator.auth0.Auth0OAuthenticator'
+c.Auth0OAuthenticator.auth0_subdomain = os.environ['AUTH0_DOMAIN']
 
-# Development authenticator
+# Development authenticator (comment out above and uncomment below to enable)
 # c.JupyterHub.authenticator_class = 'dummyauthenticator.DummyAuthenticator'
 
+# Data configuration
 data_dir = os.environ['DATA_VOLUME_CONTAINER']
 c.JupyterHub.cookie_secret_file = os.path.join(data_dir, 'jupyterhub_cookie_secret')
 c.JupyterHub.db_url = os.path.join(data_dir, 'jupyterhub.sqlite')
 
-# Whitlelist users and admins
-c.Authenticator.whitelist = whitelist = set()
-c.Authenticator.admin_users = admin = set()
+# User Configuration
+allowed_users = set()
+admin = set()
 c.JupyterHub.admin_access = True
-# comma-separated admin emails, lowercased
+# Grab THE FIRST of potentially many comma-separated admin emails, lower-cased
+# Change this when >1 admin is desired - Will 21 Jan 2022
 admin_emails = [email.strip().lower() for email in os.environ.get('ADMIN_EMAILS', '').split(',')]
```

Comment:
If there is no ADMIN_EMAILS, you probably want `[]` and not `['']`...

```suggestion
admin_emails = [email.strip().lower() for email in os.environ.get('ADMIN_EMAILS', '').split(',') if email]
```

Reply:
Technically you might want to put that in _after_ the call to `strip`, but to do that you need two loops, one inside the other.